From 9ebd3c78e1ec9cc9f8d204fd7ba2d224eba45195 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 26 Aug 2014 14:11:48 -0400
Subject: [ethernet] igb: Don't receive packets when the napi budget == 0

Message-id: <1409062359-13063-31-git-send-email-sassmann@redhat.com>
Patchwork-id: 89144
O-Subject: [RHEL7.1 PATCH 30/81] igb: Don't receive packets when the napi budget == 0
Bugzilla: 1091121
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: "Eric W. Biederman" <ebiederm@xmission.com>

Processing any incoming packets with a with a napi budget of 0
is incorrect driver behavior.

This matters as netpoll will shortly call drivers with a budget of 0
to avoid receive packet processing happening in hard irq context.

Signed-off-by: "Eric W. Biederman" <ebiederm@xmission.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 57ba34c9b068f314b219affafc19a39f8735d5e8)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/igb/igb_main.c b/drivers/net/ethernet/intel/igb/igb_main.c
index ea8b9c4..0427080 100644
--- a/drivers/net/ethernet/intel/igb/igb_main.c
+++ b/drivers/net/ethernet/intel/igb/igb_main.c
@@ -6946,7 +6946,7 @@ static bool igb_clean_rx_irq(struct igb_q_vector *q_vector, const int budget)
  unsigned int total_bytes = 0, total_packets = 0;
  u16 cleaned_count = igb_desc_unused(rx_ring);
 
- do {
+ while (likely(total_packets < budget)) {
   union e1000_adv_rx_desc *rx_desc;
 
   /* return some buffers to hardware, one at a time is too slow */
@@ -6998,7 +6998,7 @@ static bool igb_clean_rx_irq(struct igb_q_vector *q_vector, const int budget)
 
   /* update budget accounting */
   total_packets++;
- } while (likely(total_packets < budget));
+ }
 
  /* place incomplete frames back on ring for completion */
  rx_ring->skb = skb;
-- 
1.7.1